src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
     co_return co_await builder.replicate();
 }
 
+ss::future<bool>
+archival_metadata_stm::sync(model::timeout_clock::duration timeout) {
+    /*
+     * Implementation note
+     * ===================
+     *
+     * This function needs to handle two cases:
+     * Case 1: we have not synced since the last term (e.g. on leadership
+     * changes)
+     * Case 2: we have already synced (once or more) during the current
+     * term, but need to sync again (e.g. re-starting the archiver to apply
+     * topic configs)
+     *
+     * They are handled separately to the reader's delight. In theory, it is
+     * possible to fallthrough from Case 1 to Case 2. This function should be
+     * called in a loop until it succeeds.
+     */
+    if (!_raft->is_leader()) {
+        co_return false;
+    }
+
+    const auto current_term = _raft->term();
+    const auto last_applied_offset = _manifest->get_insync_offset();
+
+    if (current_term == _insync_term && _last_replicated_offset) {
+        if (last_applied_offset < _last_replicated_offset) {
+            vlog(
+              _logger.debug,
+              "Syncing archival STM from offset {} to offest in term {}",
+              last_applied_offset,
+              _last_replicated_offset);
+
+            const auto deadline = model::timeout_clock::now() + timeout;
+            const bool synced = co_await wait_no_throw(
+              *_last_replicated_offset, deadline);
+
+            if (!synced) {
+                vlog(
+                  _logger.warn,
+                  "Failed to sync archival STM from offset {} to offest {} in "
+                  "term {}",
+                  last_applied_offset,
+                  _last_replicated_offset,
+                  current_term);
+                co_return false;
+            }
+        }
```

Comment:
nit: if last_applied_offset >= _last_replicated_offset, should we co_return true immediately? Seems to reduce to the same thing, but without the below checks given there are no scheduling points with the above checks